index.html

```diff
@@ -184,6 +187,42 @@ <h3>
           </tr>
         </tbody>
       </table>
+      <h4>
```

Comment:
I think I finally understood what you meant by "header" in a previous comment.

You need to use "relevant global object's blablabla" in the algorithms, not here. Your algorithms are still broken, and these terms only make sense in the context of the algorithms, where you are given a Document `|document|` as input and which you use to derive all these other terms. Just saying "relevant global object" without retrieving it from somewhere does not work either.

What you need in this specific section is simpler:
- Drop the "The Document interface" header
- Drop this header
- Gather everything in one section just like before. The difference is that you're going to have two tables. The first one is the one that's already present in the spec; the second one is prefaced by something like _"[=Top-level traversables=] must have the following internal slots:"_. In this case, you're really referring to the _concept_ of top-level traversables, not the top-level traversable that belongs to one specific navigable.

Reply:
"Internal slot" stuff is fixed in https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/a9c074e99df1f1ef01f77e44df63d9df5d5cf459

**"_Your algorithms are still broken_"**
- Currently text in algorithm says`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`
- "_You need to use "relevant global object's blablabla" in the algorithms_" So it has already "relevant global object"?
- "_Just saying "relevant global object" without retrieving it from somewhere does not work either._" I though the  `[=navigable=]'s [=top-level traversable=]` part is the "_somewhere_", the top-level traversable of given document.

Reply:
> "Internal slot" stuff is fixed in [a9c074e](https://github.com/JuhaVainio/W3C-device-posture/commit/a9c074e99df1f1ef01f77e44df63d9df5d5cf459)
> 
> **"_Your algorithms are still broken_"**
> 
> * Currently text in algorithm says`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`

If you think in terms of code, what you need is this:

```c++
struct TopLevelTraversable {
  DevicePostureType* override;
};
```

but what you are doing is

```c++
struct TopLevelTraversable {
};

struct Document {
  DevicePostureType* override;
};

void CalculatePostureInformation(Document doc) {
  if (doc.window().navigable().topLevelTraversable().override) {
    // ...
  }
}
```

`[[PostureOverride]]` is a term that references the top-level traversable. It does not make sense to `data-dfn-for` it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.

Furthermore, while "calculate the device posture information" is almost correct, your WebDriver algorithms are not.

> * "_You need to use "relevant global object's blablabla" in the algorithms_" So it has already "relevant global object"?

What do you mean by "it" in this context?

> * "_Just saying "relevant global object" without retrieving it from somewhere does not work either._" I though the  `[=navigable=]'s [=top-level traversable=]` part is the "_somewhere_", the top-level traversable of given document.

You retrieve the navigable from the relevant global object in your algorithms, not the other way around. In any case, what I'm saying is that in this location where you're defining the slot, you don't have a "given document" like in the algorithms where you're receiving a specific Document instance.

Reply:
The code for "8.1 Calculating the device posture information" is
`If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null`
the output text is
"_If document's relevant global object's navigable's top-level traversable.[[PostureOverride]] is non-null, return it._"
Clarification: Is the outputted text OK? So only the code is incorrect?

Reply:
> [[PostureOverride]] is a term that references the top-level traversable. It does not make sense to data-dfn-for it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.

When defining PostureOverride as internal slot `[[\InternalSlot]]` style is used.
If I use `<dfn>[[\InternalSlot]]</dfn>` browser says "Internal slot _name of slot_ must be associated with WebIDL interface". In earlier comment you said "top-level traversable does not have a Web IDL definition", so which interface it should be linked? On the other hand I noticed that I do not get the error if I write something there. For example
`<dfn data-dfn-for="TopLevelTraversable">[[\PostureOverride]]</dfn>`
This is little bit weird area as the text (_TopLevelTraversable_) in not shown in spec which you actually read and it doesn't affect the created definition/link to that definition.

Reply:
>>> "You need to use "relevant global object's blablabla" in the algorithms, not here."
>>
>>So it has already "relevant global object"?
>
> What do you mean by "it" in this context?

Flow of text + my assumption
1. Your text "_You need to use "relevant global object's blablabla" in the algorithms, not here._"
2. I assumed "_need to use_" is negative term and is meant for algorithm section. Meaning text in algorithm section does not fulfill this requirement.
3. Then I would go to "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" to search if "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" already uses term "_relevant global object's_". As "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)" is already using "_relevant global object_" term I said said **it** (meaning "[8.1 Calculating the device posture information](https://github.com/JuhaVainio/W3C-device-posture/blob/d9d848ca4807f9ca82e392f6305b9631daa6272f/index.html#L478)") has already "_relevant global object's_"

Reply:
> The code for "8.1 Calculating the device posture information" is `If |document:Document|'s [=relevant global object=]'s [=navigable=]'s [=top-level traversable=].{{Document/[[PostureOverride]]}} is non-null` the output text is "_If document's relevant global object's navigable's top-level traversable.[[PostureOverride]] is non-null, return it._" Clarification: Is the outputted text OK? So only the code is incorrect?

It's unusual to use the "foo.bar" notation when "foo" is not a variable itself. A more common way of writing this would be "document's relevant global object's navigable's top-level traversable's [[PostureOverride]]". Alternatively, you can break it down into more than one step. For example:

1. Let `|topLevelTraversable|` be `|document|`'s relevant global object's navigable's top-level traversable.
2. If `|topLevelTraversable|.[[PostureOverride]]` is [...]`.

> > [[PostureOverride]] is a term that references the top-level traversable. It does not make sense to data-dfn-for it to "Document". What you need is to fix the data-dfn-for and its reference in that algorithm.
> 
> When defining PostureOverride as internal slot `[[\InternalSlot]]` style is used. If I use `<dfn>[[\InternalSlot]]</dfn>` browser says "Internal slot _name of slot_ must be associated with WebIDL interface". In earlier comment you said "top-level traversable does not have a Web IDL definition", so which interface it should be linked? On the other hand I noticed that I do not get the error if I write something there. For example `<dfn data-dfn-for="TopLevelTraversable">[[\PostureOverride]]</dfn>` This is little bit weird area as the text (_TopLevelTraversable_) in not shown in spec which you actually read and it doesn't affect the created definition/link to that definition.

The warning is a bit misleading, as you don't need to specify a Web IDL interface necessarily; any `<dfn>` works. I couldn't get `[[]]` to work with the `[= =]` shorthand syntax though; what did work is something like this: `<a data-link-for="top-level traversable">[[\PostureOverride]]</a>`.

Reply:
https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49

> any `<dfn>` works.

I assume you meant "any made up interface name". So I used "top-level traversable" in "data-dfn-for" when defining PostureOverride internal slot.

> 1. Let |topLevelTraversable|
> 2. If |topLevelTraversable|

In https://github.com/JuhaVainio/W3C-device-posture/pull/1/commits/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49 I have this two step approach.

Reply:
> [cd4e35a](https://github.com/JuhaVainio/W3C-device-posture/commit/cd4e35a8755ce033f3cbfe6d27ce467f923dbb49)
> 
> > any `<dfn>` works.
> 
> I assume you meant "any made up interface name". So I used "top-level traversable" in "data-dfn-for" when defining PostureOverride internal slot.

Well using it with a definition that does not exist does not make much sense. "top-level traversable" is a valid `<dfn>` coming from the HTML spec.